File: JavaAlgorithm/src/main/java/algo/tzashinorpu/SpecialSubject/BinaryTree/findBottomLeftValue_513.java
package algo.tzashinorpu.SpecialSubject.BinaryTree;

import algo.tzashinorpu.SpecialSubject.TreeNode;

public class findBottomLeftValue_513 {
    int res = 0;
    int level = 0;

    public int findBottomLeftValue(TreeNode root) {
        // 初始为 1 是防止只有一个节点的案例
        helper(root, 1);
        return res;
    }

    private void helper(TreeNode root, int level) {
        if (root == null) return;
        helper(root.left, level + 1);
        if (level > this.level) {
            res = root.val;
            this.level = level;
        }
        helper(root.right, level + 1);
    }
}

File: Clion/ZeroTrac/Tree/findBottomLeftValue_513.cpp
//
// Created by TzashiNorpu on 11/2/2022.
//
#include "tree.h"

using namespace Tree;

int Solution::findBottomLeftValue(TreeNode *root) {
  int max_level = 0;
  int res = 0;
  function<void(TreeNode *, int)> helper = [&](TreeNode *node, int level) -> void {
      if (!node) return;
      helper(node->left, level + 1);
      if (level > max_level) {
        max_level = level;
        res = node->val;
      }
      helper(node->right, level + 1);
  };
  helper(root, 1);
  return res;
}